# LIBRERIE

In [403]:
import pandas as pd
import os

In [404]:
os.getcwd()

'c:\\Users\\eliav\\Desktop\\Road_accidents_analysis\\Notebooks'

# CARICAMENTO DF GREZZI DI ISTAT E SITUAS e SISTEMAZIONE PER SUCCESSIVA COMBINAZIONE DEI DUE 

In [405]:
df_situas_raw = pd.read_csv("../Data/Df_raw/df_situas_raw.csv")
df_situas_raw

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD,Codice Provincia (Storico)
0,1,1,1,1001,1001,Agliè,NaN,TO,0,0,2574.0,2001,"13,1462",2001,2557,2001,2001,NaN
1,1,1,1,1002,1002,Airasca,NaN,TO,0,0,3554.0,2001,"15,7393",2001,3543,2001,2001,NaN
2,1,1,1,1003,1003,Ala di Stura,NaN,TO,0,0,479.0,2001,"46,3315",2001,480,2001,2001,NaN
3,1,1,1,1004,1004,Albiano d'Ivrea,NaN,TO,0,0,1696.0,2001,"11,7315",2001,1687,2001,2001,NaN
4,1,1,1,1005,1005,Alice Superiore,NaN,TO,0,0,616.0,2001,"7,3796",2001,619,2001,2001,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
192726,5,20,111,111103,111103,Villaputzu,NaN,SU,0,0,4509.0,2021,"181,3947",2024,4368,2024,2024,111.0
192727,5,20,111,111104,111104,Villasalto,NaN,SU,0,0,989.0,2021,"130,3596",2024,903,2024,2024,111.0
192728,5,20,111,111105,111105,Villasimius,NaN,SU,0,0,3705.0,2021,"58,1781",2024,3735,2024,2024,111.0
192729,5,20,111,111106,111106,Villasor,NaN,SU,0,0,6599.0,2021,"86,8137",2024,6534,2024,2024,111.0


In [406]:
df_situas_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 192731 entries, 0 to 192730
Data columns (total 18 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Codice Ripartizione geografica  192731 non-null  int64  
 1   Codice Regione                  192731 non-null  int64  
 2   Codice Provincia/Uts            192731 non-null  int64  
 3   Codice Comune (alfanumerico)    192731 non-null  int64  
 4   Codice Comune (numerico)        192731 non-null  int64  
 5   Comune                          192731 non-null  str    
 6   Comune (dizione straniera)      2911 non-null    str    
 7   Sigla automobilistica           190523 non-null  str    
 8   Capoluogo di Provincia/Uts      192731 non-null  int64  
 9   Capoluogo di Regione            192731 non-null  int64  
 10  Popolazione legale              192722 non-null  float64
 11  Anno Censimento                 192731 non-null  int64  
 12  Superficie (Kmq)           

In [407]:
df_situas_raw["Codice Comune (numerico)"].nunique()

8578

In [408]:
df_situas_raw= df_situas_raw[["Codice Regione", "Codice Provincia/Uts",
                      "Codice Comune (numerico)",
                      "Comune","Capoluogo di Provincia/Uts", "Capoluogo di Regione", "Superficie (Kmq)", 
                      "Popolazione residente", "Anno (Popolazione residente)", "TIME_PERIOD"]]

df_situas_raw

,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD
0,1,1,1001,Agliè,0,0,"13,1462",2557,2001,2001
1,1,1,1002,Airasca,0,0,"15,7393",3543,2001,2001
2,1,1,1003,Ala di Stura,0,0,"46,3315",480,2001,2001
3,1,1,1004,Albiano d'Ivrea,0,0,"11,7315",1687,2001,2001
4,1,1,1005,Alice Superiore,0,0,"7,3796",619,2001,2001
...,...,...,...,...,...,...,...,...,...,...
192726,20,111,111103,Villaputzu,0,0,"181,3947",4368,2024,2024
192727,20,111,111104,Villasalto,0,0,"130,3596",903,2024,2024
192728,20,111,111105,Villasimius,0,0,"58,1781",3735,2024,2024
192729,20,111,111106,Villasor,0,0,"86,8137",6534,2024,2024


In [409]:
df_situas_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 192731 entries, 0 to 192730
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype
---  ------                        --------------   -----
 0   Codice Regione                192731 non-null  int64
 1   Codice Provincia/Uts          192731 non-null  int64
 2   Codice Comune (numerico)      192731 non-null  int64
 3   Comune                        192731 non-null  str  
 4   Capoluogo di Provincia/Uts    192731 non-null  int64
 5   Capoluogo di Regione          192731 non-null  int64
 6   Superficie (Kmq)              192731 non-null  str  
 7   Popolazione residente         192731 non-null  int64
 8   Anno (Popolazione residente)  192731 non-null  int64
 9   TIME_PERIOD                   192731 non-null  int64
dtypes: int64(8), str(2)
memory usage: 14.7 MB


In [410]:
df_situas_raw[df_situas_raw["Superficie (Kmq)"].str.contains("\.", regex=True)]["Superficie (Kmq)"]#così mi fa vedere solo la colonna Superficie con i caratteri contenti .
#N.B. ricorda lo \ prima del . altrimenti cercherebbe una concatenazione di parole e non proprio il punto come valore

#regex=False -->Interpreta la stringa come testo letterale, quindi i caratteri speciali sono esattamente quello che sono 
                #a livello letterale. Serve per avere una stringa esatta
#regex=True --> quando vuoi sfruttare i simboli delle espressioni regolari (\s (spazio), \d, ^, $, ., +, *, ecc.) e 
                # non che sia letterale il carattere speciale

5002      1.287,3372
13104     1.287,3371
21204     1.287,3371
29305     1.287,3371
37406     1.287,3371
45507     1.287,3371
53608     1.287,3371
61709     1.287,3371
69707     1.287,3371
77801     1.287,3371
85893     1.287,3895
93985     1.287,3895
102076    1.287,3895
110135    1.287,3895
118183    1.287,3895
126186    1.287,3895
134169    1.287,3895
142128    1.287,3895
150049    1.287,3895
157953    1.287,3716
165854    1.288,1875
173758    1.288,1875
181658    1.288,1875
189554    1.288,1875
Name: Superficie (Kmq), dtype: str

In [411]:
df_situas_raw["Superficie (Kmq)"] = (df_situas_raw["Superficie (Kmq)"].str.replace(".", "", regex=False)
                                        .str.replace(",", ".", regex=False))

#controllo che non ci siano più , e che effettivamente sia stato tutto sostituisto dal.
df_situas_raw[df_situas_raw["Superficie (Kmq)"].str.contains("\,", regex=True)]["Superficie (Kmq)"]

Series([], Name: Superficie (Kmq), dtype: str)

In [412]:
#converto i dtype in float
df_situas_raw["Superficie (Kmq)"] = df_situas_raw["Superficie (Kmq)"].astype(float)

In [413]:
df_situas_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 192731 entries, 0 to 192730
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Codice Regione                192731 non-null  int64  
 1   Codice Provincia/Uts          192731 non-null  int64  
 2   Codice Comune (numerico)      192731 non-null  int64  
 3   Comune                        192731 non-null  str    
 4   Capoluogo di Provincia/Uts    192731 non-null  int64  
 5   Capoluogo di Regione          192731 non-null  int64  
 6   Superficie (Kmq)              192731 non-null  float64
 7   Popolazione residente         192731 non-null  int64  
 8   Anno (Popolazione residente)  192731 non-null  int64  
 9   TIME_PERIOD                   192731 non-null  int64  
dtypes: float64(1), int64(8), str(1)
memory usage: 14.7 MB


In [414]:
#controllo per vedere se per un comune sono associati 24 anni
df_situas_raw[df_situas_raw["Codice Comune (numerico)"] == 1001][["Codice Comune (numerico)", "TIME_PERIOD"]]

,Codice Comune (numerico),TIME_PERIOD
0,1001,2001
8102,1001,2002
16204,1001,2003
24304,1001,2004
32405,1001,2005
40506,1001,2006
48607,1001,2007
56708,1001,2008
64809,1001,2009
72909,1001,2010


In [415]:
#osservo per ogni codice comunale quanti anni differenti vengono registrati
df_situas_raw.groupby("Codice Comune (numerico)")["TIME_PERIOD"].nunique().value_counts()

TIME_PERIOD
24    7488
16     145
9      144
8      126
10     123
15     118
5       80
18      65
14      64
13      57
17      43
6       25
11      25
19      23
7       19
23       8
22       7
12       5
2        4
1        4
4        3
20       2
Name: count, dtype: int64

Qui si vede chiaramente che non per tutti i comuni sono stati registrati i dati per 24 anni, ci sono state variazioni nei codici / comuni

# DF ISTAT

In [416]:
df_istat_raw = pd.read_csv("../Data/Df_raw/df_istat_raw.csv")
df_istat_raw

,DATAFLOW,FREQ,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,OBS_STATUS,NOTE_DS,NOTE_REF_AREA,NOTE_DATA_TYPE,NOTE_RESULT,NOTE_TIME_PERIOD,BASE_PER,UNIT_MEAS,UNIT_MULT
0,IT1:41_983(1.0),A,1001,KILLINJ,F,2001,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,IT1:41_983(1.0),A,1001,KILLINJ,F,2002,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,IT1:41_983(1.0),A,1001,KILLINJ,F,2003,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,IT1:41_983(1.0),A,1001,KILLINJ,F,2004,13,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,IT1:41_983(1.0),A,1001,KILLINJ,F,2005,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,IT1:41_983(1.0),A,111107,ROADACC,9,2020,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573548,IT1:41_983(1.0),A,111107,ROADACC,9,2021,5,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573549,IT1:41_983(1.0),A,111107,ROADACC,9,2022,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573550,IT1:41_983(1.0),A,111107,ROADACC,9,2023,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [417]:
df_istat_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   DATAFLOW          573552 non-null  str    
 1   FREQ              573552 non-null  str    
 2   REF_AREA          573552 non-null  int64  
 3   DATA_TYPE         573552 non-null  str    
 4   RESULT            573552 non-null  str    
 5   TIME_PERIOD       573552 non-null  int64  
 6   OBS_VALUE         573552 non-null  int64  
 7   OBS_STATUS        0 non-null       float64
 8   NOTE_DS           0 non-null       float64
 9   NOTE_REF_AREA     0 non-null       float64
 10  NOTE_DATA_TYPE    0 non-null       float64
 11  NOTE_RESULT       0 non-null       float64
 12  NOTE_TIME_PERIOD  0 non-null       float64
 13  BASE_PER          0 non-null       float64
 14  UNIT_MEAS         0 non-null       float64
 15  UNIT_MULT         0 non-null       float64
dtypes: float64(9), int64(3), str(4)

Unirò i due df (istat e situas) per i codici comunali (ref_area e codice comune numerico) e i rispettivi time_period dopo aver eliminato le colonne che non mi interessano

In [418]:
df_istat_raw= df_istat_raw[["TIME_PERIOD","REF_AREA","DATA_TYPE","RESULT","OBS_VALUE"]]
df_istat_raw

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE
0,2001,1001,KILLINJ,F,10
1,2002,1001,KILLINJ,F,10
2,2003,1001,KILLINJ,F,7
3,2004,1001,KILLINJ,F,13
4,2005,1001,KILLINJ,F,2
...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2
573548,2021,111107,ROADACC,9,5
573549,2022,111107,ROADACC,9,1
573550,2023,111107,ROADACC,9,4


In [419]:
#anni registrati
df_istat_raw["TIME_PERIOD"].unique()

array([2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022,
       2023, 2024])

In [420]:
# quante osservazioni ci sono
df_istat_raw.groupby("DATA_TYPE").size()
#KILLINJ    382368 perchè unendo le osservazioni di morti e feriti fa si che le registrazioni degli incidenti (ROADACC) siano moltiplicate per 2

DATA_TYPE
KILLINJ    382368
ROADACC    191184
dtype: int64

In [421]:
# quanti eventi/persone rappresentano quelle osservazioni
df_istat_raw.groupby("RESULT")["OBS_VALUE"].sum()

RESULT
9    4779141
F    6709026
M     101944
Name: OBS_VALUE, dtype: int64

In [422]:
print(f"Gli incidenti totali dal 2001 al 2024 sono : {df_istat_raw[df_istat_raw["RESULT"] == "9"]["OBS_VALUE"].sum()}")
print(f"I feriti totali dal 2001 al 2024 sono : {df_istat_raw[df_istat_raw["RESULT"] == "F"]["OBS_VALUE"].sum()}")
print(f"I morti totali dal 2001 al 2024 sono : {df_istat_raw[df_istat_raw["RESULT"] == "M"]["OBS_VALUE"].sum()}")

Gli incidenti totali dal 2001 al 2024 sono : 4779141
I feriti totali dal 2001 al 2024 sono : 6709026
I morti totali dal 2001 al 2024 sono : 101944


In [423]:
df_istat_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype
---  ------       --------------   -----
 0   TIME_PERIOD  573552 non-null  int64
 1   REF_AREA     573552 non-null  int64
 2   DATA_TYPE    573552 non-null  str  
 3   RESULT       573552 non-null  str  
 4   OBS_VALUE    573552 non-null  int64
dtypes: int64(3), str(2)
memory usage: 21.9 MB


In [424]:
print(f"I caratteri duplicati nel dataframe situas grezzo sono : {df_istat_raw.duplicated().sum()}")
print ("="*10)
print(f"I caratteri duplicati nel dataframe istat grezzo sono : {df_istat_raw.duplicated().sum()}")

I caratteri duplicati nel dataframe situas grezzo sono : 0
I caratteri duplicati nel dataframe istat grezzo sono : 0


In [425]:
print(df_istat_raw[["REF_AREA", "TIME_PERIOD"]].dtypes)
print ("="*10)
print(df_situas_raw[["Codice Comune (numerico)" , "TIME_PERIOD"]].dtypes)

REF_AREA       int64
TIME_PERIOD    int64
dtype: object
Codice Comune (numerico)    int64
TIME_PERIOD                 int64
dtype: object


In [426]:
print(f"I codici numerici unici nel df istat sono {df_istat_raw["REF_AREA"].nunique()}")
print ("="*10)
print(f"I codici numerici unici nel df istat sono {df_situas_raw["Codice Comune (numerico)"].nunique()}")

I codici numerici unici nel df istat sono 8578
I codici numerici unici nel df istat sono 8578


In [427]:
sbatti=df_situas_raw.groupby("Codice Comune (numerico)")["Comune"].nunique()

sbatti[sbatti > 1].sum()

np.int64(72)

In [428]:
df_situas_raw.groupby("Comune")["Codice Comune (numerico)"].nunique()

Comune
Abano Terme              1
Abbadia Cerreto          1
Abbadia Lariana          1
Abbadia San Salvatore    1
Abbasanta                1
                        ..
Zuglio                   1
Zumaglia                 1
Zumpano                  1
Zungoli                  1
Zungri                   1
Name: Codice Comune (numerico), Length: 8250, dtype: int64

# MERGE & CLEANING DEI DUE DF GREZZI

In [429]:
df_final_raw= pd.merge(df_istat_raw, df_situas_raw, left_on=["REF_AREA","TIME_PERIOD"], right_on=["Codice Comune (numerico)","TIME_PERIOD"], how="left", indicator=True)

df_final_raw

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),_merge
0,2001,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2557.0,2001.0,both
1,2002,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2538.0,2002.0,both
2,2003,1001,KILLINJ,F,7,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2588.0,2003.0,both
3,2004,1001,KILLINJ,F,13,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2679.0,2004.0,both
4,2005,1001,KILLINJ,F,2,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2674.0,2005.0,both
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1937,2549.0,2020.0,both
573548,2021,111107,ROADACC,9,5,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2536.0,2021.0,both
573549,2022,111107,ROADACC,9,1,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2575.0,2022.0,both
573550,2023,111107,ROADACC,9,4,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2616.0,2023.0,both


In [430]:
#Guardo se qualcosa non è stato matchato (quindi NON BOTH)
df_final_raw["_merge"].value_counts()

_merge
both          572700
left_only        852
right_only         0
Name: count, dtype: int64

In [431]:
left_only=df_final_raw[df_final_raw["_merge"]=="left_only"]

display(left_only)
display(f"Ci sono {left_only.shape[0]} rows che non hanno trovato corrispondenza nel df situas")

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),_merge
65103,2019,6064,KILLINJ,F,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
65122,2019,6064,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
65141,2019,6064,ROADACC,9,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
66828,2019,6089,KILLINJ,F,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
66847,2019,6089,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
566295,2016,107022,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566306,2016,107022,ROADACC,9,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566317,2016,107023,KILLINJ,F,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566328,2016,107023,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only


'Ci sono 852 rows che non hanno trovato corrispondenza nel df situas'

In [432]:
# id comunali che non hanno match
left_only["REF_AREA"].nunique()

284

Numero di incidenti, feriti e morti a cui corrispondono

In [433]:
left_only.groupby("RESULT")["OBS_VALUE"].sum()

RESULT
9    6187
F    9163
M     130
Name: OBS_VALUE, dtype: int64

Quanto incidono sul totale di istat, questi dati non matchati?

In [434]:
tot = df_istat_raw.groupby("RESULT")["OBS_VALUE"].sum()
non_match = left_only.groupby("RESULT")["OBS_VALUE"].sum()

risultato = pd.DataFrame({"No match": non_match,
                          "Tot ISTAT": tot,
                          "% sul totale": (non_match / tot * 100).round(3)})

risultato

,No match,Tot ISTAT,% sul totale
RESULT,,,
9,6187,4779141,0.129
F,9163,6709026,0.137
M,130,101944,0.128


### Voglio osservare se questi comuni cambiano in qualche modo o sono informazioni mancanti 

In [435]:
#prendo codice di un comune a caso e vedo qual è il primo anno di NON corrispondenza
codice = left_only["REF_AREA"].iloc[0]
anno = left_only["TIME_PERIOD"].iloc[0]

print(codice, anno)

#guardo se esisteva in df situas --->non esisteva
df_situas_raw[(df_situas_raw["Codice Comune (numerico)"] == codice) &
                (df_situas_raw["TIME_PERIOD"] == anno)]

6064 2019


,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD


In [436]:
#guardo quel codice preso come esempio fino a che anno è stato registrato
df_situas_raw[df_situas_raw["Codice Comune (numerico)"] == codice][["Codice Comune (numerico)", "Comune", "TIME_PERIOD"]].sort_values("TIME_PERIOD")

,Codice Comune (numerico),Comune,TIME_PERIOD
920,6064,Cuccaro Monferrato,2001
9022,6064,Cuccaro Monferrato,2002
17124,6064,Cuccaro Monferrato,2003
25224,6064,Cuccaro Monferrato,2004
33325,6064,Cuccaro Monferrato,2005
41426,6064,Cuccaro Monferrato,2006
49527,6064,Cuccaro Monferrato,2007
57628,6064,Cuccaro Monferrato,2008
65729,6064,Cuccaro Monferrato,2009
73829,6064,Cuccaro Monferrato,2010


In [437]:
#cerco di capire se Cuccaro Monferrato (preso come esempio) sia stato soppresso oppure unito ad un altro comune
df_situas_raw[df_situas_raw["Comune"].str.contains("CUCCARO MONFERRATO",case=False,na=False)][
    ["Comune","Codice Comune (numerico)","TIME_PERIOD"]].sort_values("TIME_PERIOD")

,Comune,Codice Comune (numerico),TIME_PERIOD
920,Cuccaro Monferrato,6064,2001
9022,Cuccaro Monferrato,6064,2002
17124,Cuccaro Monferrato,6064,2003
25224,Cuccaro Monferrato,6064,2004
33325,Cuccaro Monferrato,6064,2005
41426,Cuccaro Monferrato,6064,2006
49527,Cuccaro Monferrato,6064,2007
57628,Cuccaro Monferrato,6064,2008
65729,Cuccaro Monferrato,6064,2009
73829,Cuccaro Monferrato,6064,2010


In [438]:

# primo_left = (left_only.groupby("REF_AREA")["TIME_PERIOD"].min().reset_index().rename(columns={
#         "REF_AREA": "Codice Comune (numerico)",
#         "TIME_PERIOD": "Primo left_only"}))

# ultimi_anni = (df_situas_raw[df_situas_raw["Codice Comune (numerico)"].isin(left_only["REF_AREA"])]
#         .groupby("Codice Comune (numerico)")["TIME_PERIOD"].max().reset_index())

# controllo = ultimi_anni.merge(primo_left,on="Codice Comune (numerico)")

# controllo["Differenza"] = (controllo["Primo left_only"] -controllo["TIME_PERIOD"])

# controllo.sort_values("Differenza")
# controllo["Differenza"].value_counts()

La merge è stata effettuata sulla chiave composta (REF_AREA, TIME_PERIOD) e validata tramite validate="many_to_one". Le osservazioni senza corrispondenza (left_only) sono risultate riconducibili quasi esclusivamente a variazioni amministrative dei comuni. Per 283 dei 284 codici comunali coinvolti, il primo anno senza corrispondenza coincide con l'anno successivo all'ultima presenza nel dataset SITUAS, indicando fusioni, soppressioni o cambi di codice. È stato individuato un solo caso anomalo, oggetto di verifica puntuale.

# CREAZIONE DF FINALE CON SOLO COLONNE DI INTERESSE

In [439]:
df_final_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 15 columns):
 #   Column                        Non-Null Count   Dtype   
---  ------                        --------------   -----   
 0   TIME_PERIOD                   573552 non-null  int64   
 1   REF_AREA                      573552 non-null  int64   
 2   DATA_TYPE                     573552 non-null  str     
 3   RESULT                        573552 non-null  str     
 4   OBS_VALUE                     573552 non-null  int64   
 5   Codice Regione                572700 non-null  float64 
 6   Codice Provincia/Uts          572700 non-null  float64 
 7   Codice Comune (numerico)      572700 non-null  float64 
 8   Comune                        572700 non-null  str     
 9   Capoluogo di Provincia/Uts    572700 non-null  float64 
 10  Capoluogo di Regione          572700 non-null  float64 
 11  Superficie (Kmq)              572700 non-null  float64 
 12  Popolazione residente         572700 non-

In [440]:
df_final_raw.isna().sum()

TIME_PERIOD                       0
REF_AREA                          0
DATA_TYPE                         0
RESULT                            0
OBS_VALUE                         0
Codice Regione                  852
Codice Provincia/Uts            852
Codice Comune (numerico)        852
Comune                          852
Capoluogo di Provincia/Uts      852
Capoluogo di Regione            852
Superficie (Kmq)                852
Popolazione residente           852
Anno (Popolazione residente)    852
_merge                            0
dtype: int64

In [441]:
df_final_clean= df_final_raw

df_final_clean

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),_merge
0,2001,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2557.0,2001.0,both
1,2002,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2538.0,2002.0,both
2,2003,1001,KILLINJ,F,7,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2588.0,2003.0,both
3,2004,1001,KILLINJ,F,13,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2679.0,2004.0,both
4,2005,1001,KILLINJ,F,2,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2674.0,2005.0,both
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1937,2549.0,2020.0,both
573548,2021,111107,ROADACC,9,5,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2536.0,2021.0,both
573549,2022,111107,ROADACC,9,1,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2575.0,2022.0,both
573550,2023,111107,ROADACC,9,4,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2616.0,2023.0,both


In [ ]:
#PRIMA DI ELIMINARE LA COLONNA MERGE MI RICREO IL DF CON SOLO QUELLI CHE SI SONO MATCHATI
#SI VANNO A PERDERE SOLO 850 RECORD CIRCA SU PIU' DI 500000 

df_final_clean=df_final_clean[df_final_clean["_merge"] == "both"].copy()

In [443]:
#Elimino merge che non mi serve più, il codice d
df_final_clean=df_final_clean.drop(columns=["_merge","Anno (Popolazione residente)","Codice Comune (numerico)"]) #ho dovuto riscrivere df_final_clean= perchè drop non eliminerebbe 
                                                            #in maniera permanente la colonna, se non volessi salvare il 
                                                            #df come un nuovo df, dovrei mettere drop(columns=["_merge"], inplace=True)

df_final_clean

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente
0,2001,1001,KILLINJ,F,10,1.0,1.0,Agliè,0.0,0.0,13.1462,2557.0
1,2002,1001,KILLINJ,F,10,1.0,1.0,Agliè,0.0,0.0,13.1462,2538.0
2,2003,1001,KILLINJ,F,7,1.0,1.0,Agliè,0.0,0.0,13.1462,2588.0
3,2004,1001,KILLINJ,F,13,1.0,1.0,Agliè,0.0,0.0,13.1462,2679.0
4,2005,1001,KILLINJ,F,2,1.0,1.0,Agliè,0.0,0.0,13.1462,2674.0
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20.0,111.0,Villaspeciosa,0.0,0.0,27.1937,2549.0
573548,2021,111107,ROADACC,9,5,20.0,111.0,Villaspeciosa,0.0,0.0,27.1943,2536.0
573549,2022,111107,ROADACC,9,1,20.0,111.0,Villaspeciosa,0.0,0.0,27.1943,2575.0
573550,2023,111107,ROADACC,9,4,20.0,111.0,Villaspeciosa,0.0,0.0,27.1943,2616.0


In [444]:
df_final_clean.isna().sum()

TIME_PERIOD                   0
REF_AREA                      0
DATA_TYPE                     0
RESULT                        0
OBS_VALUE                     0
Codice Regione                0
Codice Provincia/Uts          0
Comune                        0
Capoluogo di Provincia/Uts    0
Capoluogo di Regione          0
Superficie (Kmq)              0
Popolazione residente         0
dtype: int64

In [445]:
#controllo i tipi di dato
df_final_clean.info()

<class 'pandas.DataFrame'>
Index: 572700 entries, 0 to 573551
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   TIME_PERIOD                 572700 non-null  int64  
 1   REF_AREA                    572700 non-null  int64  
 2   DATA_TYPE                   572700 non-null  str    
 3   RESULT                      572700 non-null  str    
 4   OBS_VALUE                   572700 non-null  int64  
 5   Codice Regione              572700 non-null  float64
 6   Codice Provincia/Uts        572700 non-null  float64
 7   Comune                      572700 non-null  str    
 8   Capoluogo di Provincia/Uts  572700 non-null  float64
 9   Capoluogo di Regione        572700 non-null  float64
 10  Superficie (Kmq)            572700 non-null  float64
 11  Popolazione residente       572700 non-null  float64
dtypes: float64(6), int64(3), str(3)
memory usage: 56.8 MB


In [446]:
#SISTEMO I TIPI DELLE COLONNE

df_final_clean["Codice Regione"] = (df_final_clean["Codice Regione"].astype("Int64"))
# Int64 (con la I maiuscola) permette di avere interi e valori mancanti nella stessa colonna.
df_final_clean["Codice Provincia/Uts"] = (df_final_clean["Codice Provincia/Uts"].astype("Int64"))
df_final_clean["Capoluogo di Provincia/Uts"] = (df_final_clean["Capoluogo di Provincia/Uts"].astype("Int64"))
df_final_clean["Capoluogo di Regione"] = (df_final_clean["Capoluogo di Regione"].astype("Int64"))
df_final_clean["Popolazione residente"] = (df_final_clean["Popolazione residente"].astype("Int64"))

In [447]:
print(df_final_clean.shape) #righe, colonne
print("-"*50)
print(df_final_clean.info())

(572700, 12)
--------------------------------------------------
<class 'pandas.DataFrame'>
Index: 572700 entries, 0 to 573551
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   TIME_PERIOD                 572700 non-null  int64  
 1   REF_AREA                    572700 non-null  int64  
 2   DATA_TYPE                   572700 non-null  str    
 3   RESULT                      572700 non-null  str    
 4   OBS_VALUE                   572700 non-null  int64  
 5   Codice Regione              572700 non-null  Int64  
 6   Codice Provincia/Uts        572700 non-null  Int64  
 7   Comune                      572700 non-null  str    
 8   Capoluogo di Provincia/Uts  572700 non-null  Int64  
 9   Capoluogo di Regione        572700 non-null  Int64  
 10  Superficie (Kmq)            572700 non-null  float64
 11  Popolazione residente       572700 non-null  Int64  
dtypes: Int64(5), float64(1),

In [448]:
df_final_clean

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente
0,2001,1001,KILLINJ,F,10,1,1,Agliè,0,0,13.1462,2557
1,2002,1001,KILLINJ,F,10,1,1,Agliè,0,0,13.1462,2538
2,2003,1001,KILLINJ,F,7,1,1,Agliè,0,0,13.1462,2588
3,2004,1001,KILLINJ,F,13,1,1,Agliè,0,0,13.1462,2679
4,2005,1001,KILLINJ,F,2,1,1,Agliè,0,0,13.1462,2674
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20,111,Villaspeciosa,0,0,27.1937,2549
573548,2021,111107,ROADACC,9,5,20,111,Villaspeciosa,0,0,27.1943,2536
573549,2022,111107,ROADACC,9,1,20,111,Villaspeciosa,0,0,27.1943,2575
573550,2023,111107,ROADACC,9,4,20,111,Villaspeciosa,0,0,27.1943,2616


In [449]:
#Rinomino i valori per renderli più chiari
df_final_clean["DATA_TYPE"] = df_final_clean["DATA_TYPE"].replace({"KILLINJ": "morti_e_feriti",
                                                                    "ROADACC": "incidenti_stradali"})

df_final_clean["RESULT"] = df_final_clean["RESULT"].replace({"M": "morti",
                                                            "F": "feriti",
                                                            "9": "incidenti"})

df_final_clean=df_final_clean.rename(columns={"REF_AREA" : "ID_COMUNE", 
                                                              "Codice Provincia/Uts" : "ID_PROVINCIA",
                                                              "Codice Regione" : "ID_REGIONE",
                                                              "Capoluogo di Provincia/Uts": "CAPOLUOGO PROVINCIA",
                                                              "Capoluogo di Regione" : "CAPOLUOGO REGIONE",
                                                              "Comune" : "COMUNE",
                                                              "Popolazione residente" : "RESIDENTI",
                                                              "Superficie (Kmq)" : "SUPERFICIE (KMQ)",})
df_final_clean

,TIME_PERIOD,ID_COMUNE,DATA_TYPE,RESULT,OBS_VALUE,ID_REGIONE,ID_PROVINCIA,COMUNE,CAPOLUOGO PROVINCIA,CAPOLUOGO REGIONE,SUPERFICIE (KMQ),RESIDENTI
0,2001,1001,morti_e_feriti,feriti,10,1,1,Agliè,0,0,13.1462,2557
1,2002,1001,morti_e_feriti,feriti,10,1,1,Agliè,0,0,13.1462,2538
2,2003,1001,morti_e_feriti,feriti,7,1,1,Agliè,0,0,13.1462,2588
3,2004,1001,morti_e_feriti,feriti,13,1,1,Agliè,0,0,13.1462,2679
4,2005,1001,morti_e_feriti,feriti,2,1,1,Agliè,0,0,13.1462,2674
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,incidenti_stradali,incidenti,2,20,111,Villaspeciosa,0,0,27.1937,2549
573548,2021,111107,incidenti_stradali,incidenti,5,20,111,Villaspeciosa,0,0,27.1943,2536
573549,2022,111107,incidenti_stradali,incidenti,1,20,111,Villaspeciosa,0,0,27.1943,2575
573550,2023,111107,incidenti_stradali,incidenti,4,20,111,Villaspeciosa,0,0,27.1943,2616


In [450]:
#AGGIUNGO IL NOME DELLE REGIONI

regioni = {
    1: "Piemonte",
    2: "Valle d'Aosta",
    3: "Lombardia",
    4: "Trentino-Alto Adige",
    5: "Veneto",
    6: "Friuli-Venezia Giulia",
    7: "Liguria",
    8: "Emilia-Romagna",
    9: "Toscana",
    10: "Umbria",
    11: "Marche",
    12: "Lazio",
    13: "Abruzzo",
    14: "Molise",
    15: "Campania",
    16: "Puglia",
    17: "Basilicata",
    18: "Calabria",
    19: "Sicilia",
    20: "Sardegna"}

df_final_clean["REGIONE"] = df_final_clean["ID_REGIONE"].map(regioni)

In [451]:
df_final_clean["RESULT"].value_counts()
#Ho provato a ricontrollare anche in df_istat_raw e il rapporto è sempre questo, 1:1:1

RESULT
feriti       190900
morti        190900
incidenti    190900
Name: count, dtype: int64

In [452]:
os.getcwd()
os.listdir()

['00_FIXED_fetching.ipynb',
 '01_FIXED_cleanind_overview.ipynb',
 '01_V2_cleaning_overview.ipynb',
 '02a.EDA_&_Temporal_analysis.ipynb',
 '02b_EDA_Municipality_analysis_aggregate_01_24.ipynb',
 '03b_Test.ipynb',
 'FUNCTIONS.py',
 '__pycache__']

In [453]:
if not os.path.exists("../Data/Df_clean"):
    os.mkdir("../Data/Df_clean")
df_final_clean.to_csv("../Data/Df_clean/df_final_from_01_to_24_clean.csv",index=False)

In [454]:
df_incidenti= df_final_clean[df_final_clean["RESULT"]=="incidenti"]
df_incidenti.to_csv("../Data/Df_clean/df_incidenti_from_01_to_24.csv",index=False)

df_morti= df_final_clean[df_final_clean["RESULT"]=="morti"]
df_morti.to_csv("../Data/Df_clean/df_morti_from_01_to_24.csv",index=False)

df_feriti= df_final_clean[df_final_clean["RESULT"]=="feriti"]
df_feriti.to_csv("../Data/Df_clean/df_feriti_from_01_to_24.csv",index=False)